v3/round_trip_with_retry_backoff.go

```diff
@@ -0,0 +1,86 @@
+package heroku
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/cenkalti/backoff"
+)
+
+// net/http RoundTripper interface, a.k.a. Transport
+// https://godoc.org/net/http#RoundTripper
+type RoundTripWithRetryBackoff struct {
+	// Configuration fields for backoff.ExponentialBackOff
+	InitialIntervalSeconds int64
+	RandomizationFactor    float64
+	Multiplier             float64
+	MaxIntervalSeconds     int64
+	// After MaxElapsedTime the ExponentialBackOff stops.
+	// It never stops if MaxElapsedTime == 0.
+	MaxElapsedTimeSeconds int64
+}
+
+func (r RoundTripWithRetryBackoff) RoundTrip(req *http.Request) (*http.Response, error) {
+	var lastResponse *http.Response
+	var lastError error
+
+	retryableRoundTrip := func() error {
+		lastResponse, lastError = http.DefaultTransport.RoundTrip(req)
+		// Detect Heroku API rate limiting
+		// https://devcenter.heroku.com/articles/platform-api-reference#client-error-responses
+		if lastResponse.StatusCode == 429 {
+			return fmt.Errorf("Heroku API rate limited: 429 Too Many Requests")
+		}
+		return nil
+	}
+
+	rateLimitRetryConfig := &backoff.ExponentialBackOff{
+		Clock:               backoff.SystemClock,
+		InitialInterval:     time.Duration(int64WithDefault(r.InitialIntervalSeconds, int64(30))) * time.Second,
+		RandomizationFactor: float64WithDefault(r.RandomizationFactor, float64(0.25)),
+		Multiplier:          float64WithDefault(r.Multiplier, float64(2)),
+		MaxInterval:         time.Duration(int64WithDefault(r.MaxIntervalSeconds, int64(900))) * time.Second,
+		MaxElapsedTime:      time.Duration(int64WithDefault(r.MaxElapsedTimeSeconds, int64(0))) * time.Second,
+	}
+	rateLimitRetryConfig.Reset()
```

Comment:
Why did you need to do this? (https://godoc.org/github.com/cenkalti/backoff#ExponentialBackOff.Reset)

Reply:
I followed the way the backoff module's built-in convenience constructor works:
https://github.com/cenkalti/backoff/blob/adb73d5bf0d9237fab19ff58aebf658449e326df/exponential.go#L92

Not sure it's required, but it seemed prudent to follow the creators lead 😅

Reply:
That sounds logical to me 👍

Reply:
https://github.com/cenkalti/backoff/blob/adb73d5bf0d9237fab19ff58aebf658449e326df/exponential.go#L107-L108